### PR TITLE
Add GnuPG.ListSecretKeys, returning []SecretKeyListing

### PR DESCRIPTION
**Pull request by paulfurley**
* Add parseListSecretKeys and tests
* Add parseFingerprint and tests
* Add parseTimestamp and tests

### CODE REVIEW

File: gpgwrapper/gpgwrapper.go [hunk 1]
```diff
@@ -24,6 +26,12 @@ type GnuPG struct {
 	homeDir string
 }
 
+type SecretKeyListing struct {
```
Comment (idrysdale) [inline gpgwrapper/gpgwrapper.go]: Given we're exporting this, shall we comment it in line with the golang style guide?

File: gpgwrapper/gpgwrapper.go [hunk 2]
```diff
@@ -66,6 +74,21 @@ func (g *GnuPG) ImportArmoredKey(armoredKey string) (string, error) {
 	return output, nil
 }
```
Comment (idrysdale) [inline gpgwrapper/gpgwrapper.go]: Ditto

File: gpgwrapper/parser.go [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+		p.handleFingerprintLine(cols)
+		return
+
+	case "uid":
+		p.handleUidLine(cols)
+		return
+	}
+}
+
+func (p *listSecretKeysParser) End() {
```
Comment (idrysdale) [inline gpgwrapper/parser.go]: ... ditto.

Or, does this mean to be exported?

File: gpgwrapper/gpgwrapper_test.go [hunk 4]
```diff
@@ -99,6 +101,198 @@ func TestImportPublicKey(t *testing.T) {
...
+	gpg := GnuPG{homeDir: makeTempGnupgHome(t)}
+	gpg.ImportArmoredKey(ExamplePublicKey)
+	gpg.ImportArmoredKey(ExamplePrivateKey)
+	secretKeys, err := gpg.ListSecretKeys()
+	if err != nil {
+		t.Fatalf("error calling gpg.ListSecretKeys(): %v", err)
+	}
+
+	if len(secretKeys) != 1 {
+		t.Fatalf("expected 1 secret keys, got %d: %v", len(secretKeys), secretKeys)
```
Comment (idrysdale) [inline gpgwrapper/gpgwrapper_test.go]: 1 secret *key*

File: gpgwrapper/gpgwrapper_test.go [hunk 5]
```diff
@@ -99,6 +101,198 @@ func TestImportPublicKey(t *testing.T) {
...
+		t.Fatalf("fingerprints don't match, expected '%s', got '%s'",
+			expected.fingerprint, got.fingerprint)
+	}
+
+	if expected.created != got.created {
+		t.Fatalf("created don't match, expected '%s', got '%s'", expected.created, got.created)
+	}
+
+	if len(expected.uids) != len(got.uids) {
+		t.Fatalf("uids don't match, expected '%v', got '%v'", expected.uids, got.uids)
```
Comment (idrysdale) [inline gpgwrapper/gpgwrapper_test.go]: is it the number of uids that doesn't match, not the actual uids themselves?

File: gpgwrapper/parser.go [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+	if validity == "r" || validity == "e" || validity == "n" {
+		// Ignore this invalid primary key.
+		// https://github.com/gpg/gnupg/blob/master/doc/DETAILS#field-2---validity
+		return
+	}
+
+	createdTime, err := parseTimestamp(cols[5])
+
+	if err != nil {
+		panic(err) // TODO: improve
```
Comment (idrysdale) [inline gpgwrapper/parser.go]: Still TODO?

File: gpgwrapper/parser.go [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+
+	p.partialKey = &SecretKeyListing{
+		created: *createdTime,
+	}
+
+}
+
+func (p *listSecretKeysParser) handleFingerprintLine(cols []string) {
+	if p.partialKey == nil {
+		// We don't have a current key so either we're ignoring it
```
Comment (idrysdale) [inline gpgwrapper/parser.go]: Great comments - v clear

File: gpgwrapper/gpgwrapper.go (hunk 1 above)
Comment (paulfurley) [reply gpgwrapper/gpgwrapper.go]: Definitely, good point

File: gpgwrapper/gpgwrapper.go (hunk 2 above)
Comment (paulfurley) [reply gpgwrapper/gpgwrapper.go]: It's funny that, the *struct* isn't exported, but I kind of wanted to signify which methods you're actually supposed to use (only `PushLine` and `End`), but it doesn't really make sense exporting methods on an unexported object, I guess!

File: gpgwrapper/gpgwrapper.go [hunk 8]
```diff
@@ -24,6 +26,12 @@ type GnuPG struct {
 	homeDir string
 }
 
+type SecretKeyListing struct {
+	fingerprint string
```
Comment (paulfurley) [inline gpgwrapper/gpgwrapper.go]: Stupidly, these aren't exported.

File: gpgwrapper/gpgwrapper.go (hunk 2 above)
Comment (paulfurley) [reply gpgwrapper/gpgwrapper.go]: I've unexported and documented `PushLine` and `End` now.

File: gpgwrapper/gpgwrapper.go (hunk 1 above)
Comment (paulfurley) [reply gpgwrapper/gpgwrapper.go]: Done now.

File: gpgwrapper/gpgwrapper.go (hunk 8 above)
Comment (paulfurley) [reply gpgwrapper/gpgwrapper.go]: I've exported these now, since the struct is useless otherwise.

File: gpgwrapper/gpgwrapper_test.go (hunk 4 above)
Comment (paulfurley) [reply gpgwrapper/gpgwrapper_test.go]: Fixed.

File: gpgwrapper/gpgwrapper_test.go (hunk 5 above)
Comment (paulfurley) [reply gpgwrapper/gpgwrapper_test.go]: It's the number, but I figure the failed test output is more helpful if it shows both slices in full

**Comment by paulfurley** [comment]
There's also an escaping problem I need to fix!

File: gpgwrapper/parser.go (hunk 6 above)
Comment (paulfurley) [reply gpgwrapper/parser.go]: Fixed & test added in https://github.com/fluidkeys/fluidkeys/pull/48/commits/ef59e203fad63eb821b4dd6bd981392fdac1566f

**Comment by idrysdale** [comment]
Brilliant. That all looks 👌 

Go for it!

**Pull request merged**